pyanaconda/ui/tui/spokes/user.py

```diff
@@ -125,62 +193,93 @@ def mandatory(self):
+        if self._user_data.gecos and not self._user_data.name:
+            username = guess_username(self._user_data.gecos)
             valid, msg = check_username(username)
             if not valid:
                 self.errors.append(_("Invalid user name: %(name)s.\n%(error_message)s")
-                        % {"name": username, "error_message": msg})
+                                   % {"name": username, "error_message": msg})
             else:
-                self.args.name = guess_username(self.args.gecos)
+                self._user_data.name = guess_username(self._user_data.gecos)
 
-        self.args.groups = [g.strip() for g in self.args._groups.split(",") if g]
+        self._user_data.groups = [g.strip() for g in self._groups.split(",") if g]
 
         # Add or remove the user from wheel group
-        if self.args._admin and "wheel" not in self.args.groups:
-            self.args.groups.append("wheel")
-        elif not self.args._admin and "wheel" in self.args.groups:
-            self.args.groups.remove("wheel")
+        if self._is_admin and "wheel" not in self._user_data.groups:
+            self._user_data.groups.append("wheel")
+        elif not self._is_admin and "wheel" in self._user_data.groups:
+            self._user_data.groups.remove("wheel")
 
         # Add or remove the user from userlist as needed
-        if self.args._create and (self.args not in self.data.user.userList and self.args.name):
-            self.data.user.userList.append(self.args)
-        elif (not self.args._create) and (self.args in self.data.user.userList):
-            self.data.user.userList.remove(self.args)
+        if self._create_user and (self._user_data not in self.data.user.userList and self._user_data.name):
+            self.data.user.userList.append(self._user_data)
+        elif (not self._create_user) and (self._user_data in self.data.user.userList):
+            self.data.user.userList.remove(self._user_data)
 
         # encrypt and store password only if user entered anything; this should
         # preserve passwords set via kickstart
-        if self.args._use_password and len(self.args._password) > 0:
-            self.args.password = self.args._password
-            self.args.isCrypted = True
-            self.args.password_kickstarted = False
+        ask_password = self._input_handlers.password
+        if self._use_password and ask_password.value and len(ask_password.value) > 0:
+            self._user_data.password = ask_password.value
+            self._user_data.isCrypted = True
+            self._user_data.password_kickstarted = False
         # clear pw when user unselects to use pw
         else:
-            self.args.password = ""
-            self.args.isCrypted = False
-            self.args.password_kickstarted = False
+            self._user_data.password = ""
+            self._user_data.isCrypted = False
+            self._user_data.password_kickstarted = False
+
+
+class UserInputHandlers(object):
```

Comment:
I don't think it is necessary to have this class.

I was considering to have a universal class for something like a form, that could replace this class and the array of the input handlers, but I don't think that it is something what we need. The `AskUserInput` instances shouldn't be in a group just because they are of the same kind, they are just blocks of a user interface. In GUI, we also have separate attributes for items of the user interface.

Reply:
I'm missing `struct` in Python which is great for this use-case. You want them grouped together because they are not messed with everything else and it is easier for context help too.